service/chain/follower.go

```diff
@@ -0,0 +1,316 @@
+	}
+
+	var collIDs []flow.Identifier
+	err = f.db.View(operation.LookupPayloadGuarantees(blockID, &collIDs))
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup collections: %w", err)
+	}
+
+	guarantees := make([]*flow.CollectionGuarantee, 0, len(collIDs))
+	for _, collID := range collIDs {
+		var guarantee flow.CollectionGuarantee
+		err := f.db.View(operation.RetrieveGuarantee(collID, &guarantee))
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve guarantee (%x): %w", collID, err)
+		}
+		guarantees = append(guarantees, &guarantee)
+	}
+
+	return guarantees, nil
+}
+
+func (f *Follower) Transactions(height uint64) ([]*flow.TransactionBody, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, dps.ErrTimeout) {
+		return nil, dps.ErrTimeout
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var collIDs []flow.Identifier
+	err = f.db.View(operation.LookupPayloadGuarantees(blockID, &collIDs))
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup collections: %w", err)
+	}
+
+	var transactions []*flow.TransactionBody
+	for _, collID := range collIDs {
+		var collection flow.LightCollection
+		err := f.db.View(operation.RetrieveCollection(collID, &collection))
+		if err != nil {
+			return nil, fmt.Errorf("could not retrieve collection (%x): %w", collID, err)
+		}
+		for _, txID := range collection.Transactions {
+			var transaction flow.TransactionBody
+			err := f.db.View(operation.RetrieveTransaction(txID, &transaction))
+			if err != nil {
+				return nil, fmt.Errorf("could not retrieve transaction (%x): %w", txID, err)
+			}
+			transactions = append(transactions, &transaction)
+		}
+	}
+
+	return transactions, nil
+}
+
+func (f *Follower) Results(height uint64) ([]*flow.TransactionResult, error) {
+	blockID, err := f.block(height)
+	if errors.Is(err, dps.ErrTimeout) {
+		return nil, dps.ErrTimeout
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	var results []flow.TransactionResult
+	err = f.db.View(operation.LookupTransactionResultsByBlockID(blockID, &results))
+	if err != nil {
+		return nil, fmt.Errorf("could not lookup transaction results: %w", err)
+	}
+
+	// Convert to pointer slice for consistency.
+	var converted []*flow.TransactionResult
+	for _, result := range results {
+		result := result
+		converted = append(converted, &result)
+	}
+
+	return converted, nil
+}
+
+func (f *Follower) Seals(height uint64) ([]*flow.Seal, error) {
+
+	blockID, err := f.block(height)
+	if errors.Is(err, dps.ErrTimeout) {
+		return nil, dps.ErrTimeout
+	}
+	if err != nil {
+		return nil, fmt.Errorf("could not get block for height: %w", err)
+	}
+
+	// LookupPayloadSeals() returns the IDs of all the seals in the specified block.
+	// It should not be confused with LookupBlockSeal(), which returns the ID of the
+	// *last* payload seal found in the block.
```

Comment:
With that in mind, we could use the seals here to execute a sanity check against the execution result commit we used for each block. If there is _any_ mismatch, we should crash, as it means that an execution node is malicious, and we are not currently equipped to handle this.

Reply:
Where can we get the execution result commit for comparison? I assume it is available in the index built by the consensus follower, once the block gets finalized?